### PR TITLE
feat: add HTTP client with exponential backoff policy

### PR DESCRIPTION
**Pull request by miparnisari**
The existing HTTP client tries requests only once.

This new client will ensure that:

- server initialization won't fail if the OIDC keys endpoint is slow to respond
- functional tests won't fail if the dummy OIDC server is slow to start. E.g. https://github.com/openfga/openfga/runs/6839424195?check_suite_focus=true#step:7:28

### CODE REVIEW

File: cmd/openfga/openfga_test.go [hunk 1]
```diff
@@ -199,26 +191,12 @@ func TestBuildServerWithOidcAuthentication(t *testing.T) {
...
+	if resp.StatusCode != http.StatusOK || err != nil {
+		t.Fatalf("Failed to start service")
+	}
 
-			return nil
-		},
-		backoffPolicy,
-	)
-	require.NoError(t, err)
+	t.Log("Service is now up")
```
[the remark this answers is not in the document]
Comment (craigpastro) [reply cmd/openfga/openfga_test.go:201]: I took a look and didn't find anything. Logs may be line oriented, so I don't know if the same styling applies.

File: pkg/httpclient/httpclient.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
+package httpclient
```
[the remark this answers is not in the document]
Comment (jon-whit) [reply pkg/httpclient/httpclient.go:1]: Can we call this package `retryablehttp` instead of `httpclient`?

File: pkg/httpclient/httpclient.go [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+	internalClient http.Client
+}
+
+func NewRetryableHTTPClient() *RetryableHTTPClient {
+	return &RetryableHTTPClient{
+		internalClient: http.Client{},
+	}
+}
+
+func (client *RetryableHTTPClient) Do(req *http.Request) (*http.Response, []byte, error) {
```
Comment (jon-whit) [inline pkg/httpclient/httpclient.go:69]: Can we stick to the standard library signatures so we don't introduce a whole new (non-standard) library?

https://pkg.go.dev/net/http#Client.Do

File: pkg/httpclient/httpclient.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (craigpastro) [reply pkg/httpclient/httpclient.go]: Sorry, just seeing this now. Is there any reason we can just use [hashicorp/go-retryablehttp](https://github.com/hashicorp/go-retryablehttp)?

[the remark this answers is not in the document]
Comment (miparnisari) [reply pkg/httpclient/httpclient.go]: @craigpastro its license is MPL-2.0 and [it's forbidden](https://app.fossa.com/policies/licensing/53942)

[the remark this answers is not in the document]
Comment (craigpastro) [reply pkg/httpclient/httpclient.go]: Hmm, that's annoying. By the way, how do we log into Fossa?

[the remark this answers is not in the document]
Comment (craigpastro) [reply pkg/httpclient/httpclient.go]: I did find this: https://github.com/cncf/foundation/issues/138. No resolution as of now though.

File: pkg/retryablehttp/httpclient.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+func (rt *RetryableRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	rt.once.Do(rt.init)
+	return rt.Client.Do(req)
+}
+
+type RetryableHTTPClient struct {
+	internalClient http.Client
+}
+
+func NewRetryableHTTPClient() *RetryableHTTPClient {
```
Comment (craigpastro) [inline pkg/retryablehttp/httpclient.go:38]: Small nit: but do you mind calling this `NewClient`, or something similar, as this is already the `retryablehttp` package?

File: pkg/retryablehttp/httpclient.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+	)
+
+	// All retries failed
+	if err != nil {
+		return nil, err
+	}
+
+	return resp, nil
+}
+
```
Comment (craigpastro) [inline pkg/retryablehttp/httpclient.go:70]: Would you mind adding a `Get` convenience function? 
```
func (c * RetryableHTTPClient) Get(url string) (*http.Response, error) {
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return nil, err
	}
	return c.Do(req)
}
```

File: pkg/retryablehttp/httpclient.go [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+// RoundTrip executes a single HTTP transaction, but does not attempt to read the response, modify it, or close the body.
+func (rt *RetryableRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	rt.once.Do(rt.init)
+	return rt.Client.Do(req)
+}
+
+type RetryableHTTPClient struct {
+	internalClient http.Client
```
Comment (craigpastro) [inline pkg/retryablehttp/httpclient.go:35]: I want to make a retryable client that has its own Transport (for TLS). One way to do this is to use `NewClient` and then set `internalClient`. I am doing that here https://github.com/openfga/openfga/pull/2/files#diff-0b4e82b4c716948a5e14e84e6695d4fd755a5ea47aa0742573dce8fdd2b20bb4R218. In order to do this here I think I will need this to be public. Is that okay?

Comment (miparnisari) [reply pkg/retryablehttp/httpclient.go:35]: I added a new constructor

File: server/authentication/oidc/oidc.go [hunk 7]
```diff
@@ -33,8 +36,9 @@ var _ authentication.OidcAuthenticator = (*RemoteOidcAuthenticator)(nil)
 
 func NewRemoteOidcAuthenticator(issuerURL, audience string) (*RemoteOidcAuthenticator, error) {
 	oidc := &RemoteOidcAuthenticator{
-		IssuerURL: issuerURL,
-		Audience:  audience,
+		IssuerURL:  issuerURL,
+		Audience:   audience,
+		httpClient: retryablehttp.RetryableHTTPClient{},
```
Comment (elbuo8) [inline server/authentication/oidc/oidc.go:41]: And if I want to inject my own?

Comment (miparnisari) [reply server/authentication/oidc/oidc.go:41]: ehmm why would you want to?

**Comment by jon-whit** [comment]
@miparnisari After thinking about this a little more. Would it be possible to not introduce a whole new HTTP client library with retries and just use the `backoff.Retry(...)` approach to wait for the mock OIDC server to start up, just like we do with the OpenFGA server `ensureServiceUp`.

I think it'd be nice if we could avoid introducing all of the HTTP client code into the OpenFGA package just to wait for an HTTP server's healthiness. If we're going to introduce it though, it would at least be nice to put it in an `openfga/pkg` repository that has shared packages like this so we can share and reuse it elsewhere. If we're going to introduce it into our software maintenance cycle, we may as well be able to share it and evolve across other projects in the OpenFGA organization on Github.

**Comment by miparnisari** [comment]
> I think it'd be nice if we could avoid introducing all of the HTTP client code into the OpenFGA package just to wait for an HTTP server's healthiness.

That's not the only use case, I also want to retry when fetching the JWKs in the OIDC authenticator.

> it would at least be nice to put it in an `openfga/pkg` repository that has shared packages like this so we can share and reuse it elsewhere

I'm confused, we don't need to put it in a brand new repository to reuse it, no?

**Comment by jon-whit** [comment]
> I'm confused, we don't need to put it in a brand new repository to reuse it, no?

You're right, we don't have to. But for a shared library that expands beyond the scope of just this project it would make sense for it to live elsewhere. An import from openfga just for a shared retry http client is what would seem odd to me. I would expect an import like that to come from a shared repository, not openfga. That's what I meant.

File: server/authentication/oidc/oidc.go [hunk 8]
```diff
@@ -22,6 +23,8 @@ type RemoteOidcAuthenticator struct {
 
 	JwksURI string
 	JWKs    *keyfunc.JWKS
+
+	httpClient retryablehttp.RetryableHTTPClient
```
Comment (jon-whit) [inline server/authentication/oidc/oidc.go:27]: This should be `*http.Client` and we can inject it with

`retryClient.StandardClient()`

**Comment by miparnisari** [comment]
> > I'm confused, we don't need to put it in a brand new repository to reuse it, no?
> 
> You're right, we don't have to. But for a shared library that expands beyond the scope of just this project it would make sense for it to live elsewhere. An import from openfga just for a shared retry http client is what would seem odd to me. I would expect an import like that to come from a shared repository, not openfga. That's what I meant.

👍 i think when we have more things that we want to reuse we can create a new repo

File: server/authentication/oidc/oidc.go [hunk 9]
```diff
@@ -138,21 +142,21 @@ func (oidc *RemoteOidcAuthenticator) GetConfiguration() (*authentication.OidcCon
...
-		return nil, errors.Errorf("unexpected status code getting OIDC: %v", res.StatusCode)
-	}
-
 	defer res.Body.Close()
 	body, err := ioutil.ReadAll(res.Body)
 	if err != nil {
 		return nil, errors.Errorf("error reading response body: %v", err)
 	}
 
+	if res.StatusCode != http.StatusOK {
```
Comment (elbuo8) [inline server/authentication/oidc/oidc.go:156]: Why was this moved 🤔

Comment (miparnisari) [reply server/authentication/oidc/oidc.go]: Reverted this change :D

**Review by elbuo8 — APPROVED** [review]

**Review by jon-whit — APPROVED** [review]
